VPNKit protocol and Go API update

MagnusS opened the pull request:
This updates the VPNKit protocol to version 22. This change is incompatible with earlier versions of VPNKit and depends on moby/vpnkit#277.

It was previously possible to encode an IP address in the UUID to request a fixed address from VPNKit. This replaces this functionality with a `preferred_ipv4=` option that can be passed to the driver. IPs encoded in UUIDs are now ignored and treated as a regular UUID.

The new server error messages allows the driver to inform the user about some common VPNKit error conditions, such as if the IP is unavailable. The old driver would just fail with an unspecified vif error.

rn commented:
Is the new VPNKit already shipping with Docker for Mac? Should we only merge this once it is at least in `-edge`?

MagnusS commented:
@rn not yet, as we need an updated `hyperkit` binary as well. I can use the binary from this PR if necessary though. If we merge you wouldn't be able to run `hyperkit` master with `vpnkit` without updating that from master as well, but tagged releases of `vpnkit`/`hyperkit` could still be used with stable/edge. We should probably wait with updating the go api in `linuxkit` though, as the `--ip` option would stop working until there's a new `hyperkit` in D4M.

rn commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -324,21 +343,60 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
+	} else { // preferred_ipv4
+		struct msg_preferred_ipv4 cmd_preferred_ipv4 = {
+			.command = CMD_PREFERRED_IPV4,
+			.ip = OSSwapInt32(*preferred_ipv4), // VPNKit uses LE, our IP is in network byte order
```
Is VPNKit explicitly in LE or is it in host byte order. If the latter this should be `htonl` or equiv

MagnusS replied:
VPNKit explicitly uses LE

ijc commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -324,21 +343,60 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
 
-	if (really_read(fd, (uint8_t*)vif, sizeof(*vif)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to read vif info\n");
+	if (preferred_ipv4 == NULL) { // ethernet
```
TBH I'm not really sure what this comment is supposed to tell me, regardless of where it is placed.

I'd prefer if the if/else was inverted so it was `preferred_ipv4 != NULL { do stuff using preferred_ipv4 } else { do the other thing }`. With that I also think there would be no need for a comment for the `preferred_ipv4 != NULL` case since it would be self explanatory, the `else` case would have a `/* If no preferred ipv4 address configured then we «explain the fallback» */`

MagnusS replied:
Hm, yes - I agree that would be more readable, I'll swap them around and remove the pointer as suggested below

ijc commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -324,21 +343,60 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
+			char buf[257];
+			memcpy(&buf[0], &(reason->msg[0]), reason->len);
+			buf[reason->len] = '\0';
+			fprintf(stderr, "virtio-net-vpnkit: server disconnected: %s\n", buf);
```
I wonder if this works:
```
fprintf(stderr, "virtio-net-vpnkit: server disconnected: %*s\n", reason->len, (casting stuff)reason->msg)
```
?

If that doesn't work then I'm a little bit unhappy with the hardcoded assumption about the range of `reason->len` being encoded in that `char buf[257]` and hence the overflow check being implicit and (more importantly) based on code far away. An `assert(sizeof(reason->len) == 1)` would be one approach, or something like `char buf[sizeof(reason->msg)+1]` might work?

MagnusS replied:
That worked! I'll remove the buffer

ijc commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -423,7 +493,7 @@ vpnkit_create(struct pci_vtnet_softc *sc, const char *opts)
 		}
 
-		if (vpnkit_connect(fd, uuid_string, &state->vif) == 0)
+		if (vpnkit_connect(fd, uuid_string, &state->vif, (ipv4 == NULL ? NULL : &preferred_ipv4.s_addr)) == 0)
```
Perhaps you could arrange for `s_addr` to be zero being the default/unconfigured action, which would avoid the conditionals and pointer argument here, i.e. vpnkit_connect could just take a plain `in_addr_t`.

MagnusS commented:
@rn @ijc thanks for the reviews! I've pushed an update.

Re timing, I think we could merge this when we are ready to include it in D4D master/nightly (tests are green, 17.07 is released etc), which is hopefully by end of this week. We would then have to wait a few weeks with updating the go vendoring in linuxkit though (or update and break`--ip` temporarily, unless you run the nightly build). Do you think what would work?

rn commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -93,13 +95,21 @@ struct msg_init {
+	uint32_t unused;
+} __packed;
+
+struct msg_preferred_ipv4 {
```
Are the messages always the same size? Looks like `msg_ethernet` and `msg_preferred_ipv4` are. If so, a `union` will be a better structure for this.

rn commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -324,19 +343,56 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
+			reason = (struct disconnect_reason*)&(reply.data);
+			fprintf(stderr, "virtio-net-vpnkit: server disconnected: %*s\n", reason->len, reason->msg);
+			return -1;
+		case RESP_VIF: 
```
I would move this to the top as this is a the normal/common case which indicates success. `RESP_DISCONNECT` and "unknown" are both error cases and should be grouped together

rn commented on src/lib/pci_virtio_net_vpnkit.c:
```diff
@@ -108,6 +118,15 @@ struct vif_info {
+	char msg[256];
+} __packed;
+
+struct msg_response {
```
I would move this declaration above `disconnect_reason` as this is what we always get. and then `disconnect_reason` is would be in data.

Also, it might make sense to make this a union where it's either `data` or `disconnect_reason` with `response_type` the selector.

MagnusS commented:
@rn thanks - the struct's are the same size, so I updated to use unions and moved the case. I left the extra commit for review, but squashed and force pushed now.

MagnusS commented:
Merging this now as all tests passed on the build from this PR in D4D. A D4D build with updated vpnkit/hyperkit should be available soon.

MagnusS commented:
This feature is going into 17.09, so it will be in the edge rc1 release today/tomorrow. I'll make a PR to update linuxkit

rn commented:
I don't know how many linuxkit/hyperkit users we have, but possibly a few. What's the error when running linuxkit with these updates against an incompatible DfM? is it sensible? Is there a simple way to fall back to the old protocol (i had a quick look and it didn't seem to be)?

MagnusS commented:
If you run an updated linuxkit + an older DfM you will get an `invalid option` error if you use `--ip`, but it should otherwise work. 

If you run an outdated linuxkit with an updated DfM release, the `--ip` parameter will be silently ignored as the ip is not parsed from the UUID anymore.
